Refactor: picked out and cleaned up code surrounding reading and writing of game save

ivan-mogilko opened the pull request:
This picks out data processing that precedes and follows writing and restoring game data of the game's save. General purpose is to separate complemetary tasks from reading and writing actual data. Currently they are spread all around reading/writing code.
These tasks include:
* Preparing game for writing;
* Opening savegame file and checking its header information (format, description);
* Preparing game for reading (releasing particular runtime objects mostly);
* Preparing game for running again after reading save (recreating runtime objects with new data, loading new room, and so on).

Additionally introduced savegame error enumeration and made reading function return those error codes instead of quitting the game.
Merged several different pieces of code into common function that deduces what to do if restoration failed, depending on how far engine went (display on-screen message or quit).

rofl0r commented on Engine/game/savegame.h:
```diff
@@ -0,0 +1,202 @@
+#if defined (_MSC_VER)
+#include <memory>
+#elif defined (__GNUC__)
+#include <tr1/memory>
```
isn't this pulling in a requirement of stdc++11 support ?

ivan-mogilko replied:
no, it is std::tr1 namespace, which we are using for a about year in the new version. Should compile with latest gcc 4.

rofl0r replied:
ok, thanks for clarification.

gurok commented:
I understand that this has no milestone, but are you releasing 3.4.0 before or after this gets merged?

ivan-mogilko commented:
3.4.0 is secured in its corresponding branch now, release-3.4.0, so it is safe to put completely new stuff to master now. This is where our development continues from now on.

As for 3.4.0, any hotfixes should go to release-3.4.0. I will add final release tag at that branch too when it's time.

gurok commented:
Ah, okay, sorry. I missed the fact that a branch had been made.

ivan-mogilko commented:
Hmm, I maybe need to find a way to announce such things more explicitly on github.

sonneveld commented on Common/util/string_utils.cpp:
```diff
@@ -203,5 +204,12 @@ void StrUtil::WriteString(const String &s, Stream *out)
 
+void StrUtil::WriteCStr(const String &s, Stream *out, int max_len)
+{
+    max_len = Math::Min(max_len, s.GetLength());
```
Instead of silently truncating, would it be better to raise an error if it doesn't fix within max_len?

ivan-mogilko replied:
I can make it return actually written length probably?

This function is meant to forcefully truncate string, it emulates how AGS was writing limited fields originally, so it is not like an erroneous situation.

ivan-mogilko replied:
Well, I just found a bug in this function... parameter's default value -1 will make nothing to be written, instead of whole string, as was probably planned.

sonneveld replied:
hah, didn't even notice that.  Clearly more tests needed :)  Returning the written size (ignoring null char so you can compare with strlen) would be a good start I think.

sonneveld commented on Engine/ac/game.cpp:
```diff
@@ -2443,177 +2323,134 @@ int restore_game_data (Stream *in, const char *nametouse, SavedGameVersion svg_v
 }
 
-int restore_game_data (Common::Stream *in, const char *nametouse)
+bool restore_game_data(Common::Stream *in, const char *nametouse)
```
Are we converting all methods from errorcodes to bools?  Bools don't really let you have more than one error state :)

ivan-mogilko replied:
Ok, I will change it to errcode.

ivan-mogilko replied:
Ok, I will change to errcode.

ivan-mogilko replied:
This function is supposed to be called only in replay loading code that is factually not used in the engine write now. That code is not interested in precise error. Also, I do not want to expose savegame error codes to other modules, because they don't need to know it. Making this properly will require reorganizing code more, which is not worthy for this issue.
So I will leave this function having just a bool; probably rename it to indicate that it is used just for very specific case.

ivan-mogilko replied:
Ok, nevermind, I found a better way.

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+extern Bitmap *dynamicallyCreatedSurfaces[MAX_DYNAMIC_SURFACES];
+extern Bitmap *raw_saved_screen;
+extern RoomStatus troom;
+extern RoomStatus *croom;
```
Why do we still use externs for globals?

ivan-mogilko replied:
Okay, I will move these into some headers.

sonneveld replied:
Actually, could these globals be passed in as references to the save functions?  It would make it more explicit that we're changing global state. (maybe a different pull request)

It's messy but I like that we're being explicit at the top which globals we're using.

ivan-mogilko replied:
Loigcally they have to be gathered in a class, or few classes referencing each other. If I start to pass references to every object I use in savegame function now, I will end up with dozens of references.

sonneveld replied:
hrm you're right.  Perhaps session classes for different lifetimes, an application session, a per game session, etc.  Sorry, not suggesting you do that in this story. :)

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+    // ensure that the current cursor is locked
+    spriteset.precache(game.mcurs[r_data.CursorID].pic);
+
+#if (ALLEGRO_DATE > 19990103)
```
Are we still supporting allegro versions from before 1999? :)

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+
+    // If there were synced audio tracks, the time taken to load in the
+    // different channels will have thrown them out of sync, so re-time it
+    for (int i = 0; i <= MAX_SOUND_CHANNELS; ++i)
```
did you mean to iterate from 0 .. MAX_SOUND_CHANNELS?

ivan-mogilko replied:
I did not mean anything, I plain moved this code from another place.

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+        }
+    }
+
+    for (int i = 1; i < MAX_SOUND_CHANNELS; ++i)
```
i.e, this is now 1 .. MAX_SOUND_CHANNELS

ivan-mogilko replied:
These are iterations over different arrays. They may have different sizes too. I cannot tell right away why it is done so.

sonneveld replied:
Would it be worth adding a couple of comments as a warning for the next guy? Maybe it's fine.

ivan-mogilko replied:
I would, but what do I write in these comments?
This code needs investigation, because I do not remember (and possibly never knew) why does it look this way.
I honestly cannot even say if that is not a bug. I can only check that it was originally same way.

sonneveld replied:
How about "TODO: beware this strange loop range" ?  Or do we just track it  as a potential bug in youtrack?

ivan-mogilko replied:
Ok, a note to investigate later may be a good idea.

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+        unexport_gui_controls(i);
+    }
+
+    for (int i = 0; i <= MAX_SOUND_CHANNELS; ++i)
```
Another instance of 0 .. MAX_SOUND_CHANNELS

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+    const int cf_out_chan = play.crossfading_out_channel;
+    play.crossfading_in_channel = 0;
+    play.crossfading_out_channel = 0;
+    for (int i = 0; i <= MAX_SOUND_CHANNELS; ++i)
```
0 .. MAX_SOUND_CHANNELS

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -0,0 +1,596 @@
+    is_text_overlay = 0;
+
+    // cleanup dynamic sprites
+    for (int i = 1; i < spriteset.elements; ++i)
```
Ok, last for loop to comment on :)  is sprite 0 special case? I can't remember.

ivan-mogilko replied:
If I remember correctly sprite 0 was reserved to keep "invalid sprite" graphic. I do not remember how it is used though.

sonneveld commented:
Reviewed. Looks good. Mostly straightforward extraction I think?

ivan-mogilko commented:
This needs additional fixing after String's pull request was merged.

ivan-mogilko commented:
Rewrote branch.
Following @sonneveld 's comment https://github.com/adventuregamestudio/ags/pull/344#discussion_r73779708 I finally decided to refactor more, and hide couple of structs and functions that are only used in a proper sequence.

Regarding WriteCStr, I investigated bit more and realized I did a mistaske by truncating the saved field at the time of writing, while originally it was supposed to be truncated at the time of *reading*, simply because the input buffer was limited. Now when this code uses dynamic strings, this is no longer required, so I changed that and removed constants defining lengths of those fields.

ivan-mogilko commented:
Had to fix one oversight, plus added little more refactoring.

Tested saving and loading games with this branch, restoring saves made by older engine and restoring saves made by this branch in older engine. Now this pull request is done, I think.
